pdns/pkcs11signers.cc

```diff
@@ -789,33 +790,31 @@ std::string PKCS11DNSCryptoKeyEngine::hash(const std::string& msg) const {
     L<<Logger::Error<<"Could not digest using PKCS#11 token - using software workaround"<<endl;
     // FINE! I'll do this myself, then, shall I?
     switch(d_algorithm) {
-    case 5: {
-      SHA1Summer sha;
-      sha.feed(msg);
-      return sha.get();
+    case 5:
+      md_info = md_info_from_type(POLARSSL_MD_SHA1);
+      break;
+    case 8:
+      md_info = md_info_from_type(POLARSSL_MD_SHA256);
+      break;
+    case 10:
+      md_info = md_info_from_type(POLARSSL_MD_SHA512);
+      break;
+    case 13:
+      md_info = md_info_from_type(POLARSSL_MD_SHA256);
+      break;
+    case 14:
+      md_info = md_info_from_type(POLARSSL_MD_SHA384);
+      break;
     }
-    case 8: {
-      SHA256Summer sha;
-      sha.feed(msg);
-      return sha.get();
-    }
-    case 10: {
-      SHA512Summer sha;
-      sha.feed(msg);
-      return sha.get();
-    }
-    case 13: {
-      SHA256Summer sha;
-      sha.feed(msg);
-      return sha.get();
-    }
-    case 14: {
-      SHA384Summer sha;
-      sha.feed(msg);
-      return sha.get();
-    }
-    };
-  };
+
+    unsigned char hash[64];
+    md_init(&md_ctx);
+    md_init_ctx(&md_ctx, md_info);
```

Comment:
md_info ends up uninitialized if some weird algorithm is used. Maybe add a default case in select?